working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+class hierarchies. In the previous example:
+
+```dart
+printGeometry(Geometry geometry) {
+  match (geometry) {
+    case Point(var x, var y) => print("Point $y, $y");
+    case Circle(var x, var y, var radius) => print("Circle $x, $y, $radius");
+  }
+}
+```
+
+Here, `Point()` is a extractor defined something like:
+
+```dart
+class Point extends Shape {
+  static Point? extract(Shape shape) => shape is Point ? shape : null;
+}
+```
+
+**TODO: Is it necessary to define explicit extractors for downcasts? Should we
+consider allowing `is` before a extractor to do a cast check?**
+
+### Resolving extractor functions
+
+We resolve the identifier in a extractor pattern to a function like so:
+
+1.  If the name is unqualified and there is a static method defined on the
+    value's static type with that name, the extractor is that method. *This lets
+    a class define its own "named extractors" that take precedence over the
+    lexical scope as in the `hsv()` example. Making these instance or extension
+    members on the class lets them have short names without risk of top level
+    collision.*
+
+3.  Otherwise, if the name resolves to a top level function, that function is
+    the extractor. *This lets you define free-floating extractor functions.*
+
+4.  Otherwise, if the name resolves to a class and the class defines a static
+    method named `extract()`, the extractor is that method. *This lets you use
+    the name of an unrelated class as if the class itself were an extractor.
+    It's the extractor analogue of an unnamed constructor.*
+
+    **TODO: Is `extract()` the name we want?**
+
+5.  Otherwise, no extractor could be found and a compile-time error is reported.
+
+**TODO: Is this too complex?**
+
+It is a compile-time error if the resolved extractor function takes anything but
+a single mandatory positional argument. It is a compile-time error if the
+function's return type is `void`.
+
+### Single-value extractors
+
+If an extractor converts the matched value to something containing multiple
+fields, it's natural to return a tuple which conveniently implements the proper
+`field<n>` fields. But if the extractor produces only a single value, what does
+it return? There are no one-element tuples. We could provide some wrapper class
+like:
+
+```dart
+class Box<T> {
+  final T field0;
+}
+```
+
+That would enable extractor patterns like `foo(x)` to work by having `foo()`
+return a `Box<T>` wrapping the value it wants to expose. But that's kind of
+gross. A cleaner solution might be to add this to the core library:
+
+```dart
+extension Monuple<T> on T {
+  T get field0 => this;
+}
+```
+
+Then every object is implicitly also a single-element tuple.
+
+### Tuple matcher
+
+```
+tupleMatcher ::= "(" matcherFields ")"
+
+matcherFields ::= matcherField ( "," matcherField )* ","?
+matcherField ::= ( identifier ":" )? matcher
+```
+
+Destructures fields from an object. It is a compile-time error if the static
```

Comment:
So as mentioned above, does the tuple `(1, 2, 3)` match against `(a , b)`?  That's a bit surprising, but maybe ok?